src/serac/physics/tests/solid_dynamics_patch.cpp

```diff
@@ -0,0 +1,438 @@
+ *
+ * @tparam dim number of spatial dimensions
+ */
+template <int dim>
+class AffineSolution {
+public:
+  AffineSolution() : disp_grad_rate(dim), initial_displacement(dim)
+  {
+    for (int i = 0; i < dim; i++) {
+      initial_displacement(i) = b[i];
+      for (int j = 0; j < dim; j++) {
+        disp_grad_rate(i, j) = A[i][j];
+      }
+    }
+  };
+
+  /**
+   * @brief MFEM-style coefficient function corresponding to this solution
+   *
+   * @param X Coordinates of point in reference configuration at which solution is sought
+   * @param u Exact solution evaluated at \p X
+   */
+  void operator()(const mfem::Vector& X, double t, mfem::Vector& u) const
+  {
+    disp_grad_rate.Mult(X, u);
+    u *= t;
+    u += initial_displacement;
+  }
+
+  void velocity(const mfem::Vector& X, double /* t */, mfem::Vector& v) const { disp_grad_rate.Mult(X, v); }
+
+  /**
+   * @brief Apply forcing that should produce this exact displacement
+   *
+   * Given the physics module, apply boundary conditions and a source
+   * term that are consistent with the exact solution. This is
+   * independent of the domain. The solution is imposed as an essential
+   * boundary condition on the parts of the boundary identified by \p
+   * essential_boundaries. On the complement of
+   * \p essential_boundaries, the traction corresponding to the exact
+   * solution is applied.
+   *
+   * @tparam p Polynomial degree of the finite element approximation
+   * @tparam Material Type of the material model used in the problem
+   *
+   * @param material Material model used in the problem
+   * @param solid The SolidMechanics module for the problem
+   * @param essential_boundaries Boundary attributes on which essential boundary conditions are desired
+   */
+  template <int p, typename Material>
+  void applyLoads(const Material& material, SolidMechanics<p, dim>& solid, std::set<int> essential_boundaries) const
+  {
+    // essential BCs
+    auto ebc_func = [*this](const auto& X, double t, auto& u) { this->operator()(X, t, u); };
+    solid.setDisplacementBCs(essential_boundaries, ebc_func);
+
+    // natural BCs
+    auto Hdot     = make_tensor<dim, dim>([&](int i, int j) { return disp_grad_rate(i, j); });
+    auto traction = [material, Hdot](auto, auto n0, auto t) {
+      auto                     H = Hdot * t;
+      typename Material::State state;  // needs to be reconfigured for mats with state
+      tensor<double, dim, dim> sigma = material(state, H);
+      auto                     F     = Identity<dim>() + H;
+      auto                     J     = det(F);
+      auto                     P     = J * dot(sigma, inv(transpose(F)));
+      // We don't have a good way to restrict the tractions to the
+      // complement of the essential boundary segments.
+      // The following matches the case when the top and left surfaces
+      // have essential boundary conditions:
+      //
+      // auto T = (n0[0] > 0.99 || n0[1] < -0.99)? dot(P, n0) : 0.0*n0;
```

Comment:
That is true. But really, neither of those are the right way. Since this is just commented debug output, I prefer to leave it until we replace it with submesh boundary sets.

Reply:
Speaking of, we should have all the pieces to do that now.